hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/WanProtocolNegotiationOperation.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.cluster.impl.operations;
+
+import com.hazelcast.config.GroupConfig;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.spi.impl.operationservice.Operation;
+import com.hazelcast.version.MemberVersion;
+import com.hazelcast.version.Version;
+import com.hazelcast.wan.impl.WanDataSerializerHook;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.hazelcast.internal.cluster.impl.operations.WanProtocolNegotiationStatus.GROUP_NAME_MISMATCH;
+import static com.hazelcast.internal.cluster.impl.operations.WanProtocolNegotiationStatus.OK;
+import static com.hazelcast.internal.cluster.impl.operations.WanProtocolNegotiationStatus.PROTOCOL_MISMATCH;
+
+/**
+ * WAN protocol negotiation operation sent from source to target cluster to
+ * negotiate WAN protocol and return metadata about target member and
+ * cluster.
+ */
+public class WanProtocolNegotiationOperation extends Operation implements JoinOperation {
```

Comment:
Can't we move this operation, response, status and `WanReplicationService#getSupportedWanProtocolVersions()` to the EE codebase? These feel EE-specific and it looks like nothing is using them from OS.

Reply:
Good point.

Reply:
Moved `WanProtocolNegotiationStatus`, `WanProtocolNegotiationResponse` and `WanProtocolNegotiationOperation`. Not sure how to move the service method, though. It's the same story as with other EE features (`consistencyCheck`, `syncAllMaps`, ...). We have a single `WanReplicationService` interface and the OS version which does nothing (or throws an exception) and the EE version which implements the feature. I have to think about it. Maybe a better place for `getSupportedWanProtocolVersions` would be on some service provider, like the research PRD suggests. I'll think about it and address in a follow-up PR.